pkg/resource_auditlog.go

```diff
@@ -0,0 +1,237 @@
+		client.log.Error().Err(err).Msg("Failed to create auditlog")
+		return diag.FromErr(err)
+	}
+
+	if v, ok := d.GetOk(auditLogIsDefaultFieldName); ok {
+		isDefault := v.(bool)
+		if isDefault {
+			if _, err := auditc.SetDefaultAuditLog(client.ctxWithToken, &audit.SetDefaultAuditLogRequest{
+				OrganizationId: result.GetOrganizationId(),
+				AuditlogId:     result.GetId(),
+			}); err != nil {
+				return diag.FromErr(err)
+			}
+		}
+	}
+
+	if result != nil {
+		d.SetId(result.Id)
+	}
+	return resourceAuditLogRead(ctx, d, m)
+}
+
+// expandAuditLogResource will take a Terraform flat map schema data and turn it into an Oasis Audit Log.
+func expandAuditLogResource(d *schema.ResourceData) (*audit.AuditLog, error) {
+	ret := &audit.AuditLog{}
+	if v, ok := d.GetOk(auditLogNameFieldName); ok {
+		ret.Name = v.(string)
+	} else {
+		return nil, fmt.Errorf("unable to find parse field %s", auditLogNameFieldName)
+	}
+	if v, ok := d.GetOk(auditLogDescriptionFieldName); ok {
+		ret.Description = v.(string)
+	}
+	if v, ok := d.GetOk(auditLogOrganizationFieldName); ok {
+		ret.OrganizationId = v.(string)
+	} else {
+		return nil, fmt.Errorf("unable to find parse field %s", auditLogOrganizationFieldName)
+	}
+
+	ret.Destinations = []*audit.AuditLog_Destination{{Type: audit.DestinationCloud}}
+
+	return ret, nil
+}
+
+// resourceAuditLogDelete will delete a given AuditLog resource based on the given ID
+func resourceAuditLogDelete(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	client := m.(*Client)
+	if err := client.Connect(); err != nil {
+		client.log.Error().Err(err).Msg("Failed to connect to api")
+		return diag.FromErr(err)
+	}
+
+	auditc := audit.NewAuditServiceClient(client.conn)
+	if _, err := auditc.DeleteAuditLog(client.ctxWithToken, &common.IDOptions{Id: d.Id()}); err != nil {
+		client.log.Error().Err(err).Str("auditlog-id", d.Id()).Msg("Failed to delete Audit Log")
+		return diag.FromErr(err)
+	}
+	d.SetId("") // called automatically, but added to be explicit
+	return nil
+}
+
+// resourceAuditLogUpdate will take a resource diff and apply changes accordingly if there are any.
+func resourceAuditLogUpdate(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	client := m.(*Client)
+	if err := client.Connect(); err != nil {
+		client.log.Error().Err(err).Msg("Failed to connect to api")
+		return diag.FromErr(err)
+	}
+
+	auditc := audit.NewAuditServiceClient(client.conn)
+	auditLog, err := auditc.GetAuditLog(client.ctxWithToken, &common.IDOptions{Id: d.Id()})
+	if err != nil {
+		client.log.Error().Err(err).Msg("Failed to find AuditLog")
+		d.SetId("")
+		return diag.FromErr(err)
+	}
+	isChangedDefault := false
+	newDefaultAuditLogID := ""
+	// Main fields
+	if d.HasChange(auditLogNameFieldName) {
+		auditLog.Name = d.Get(auditLogNameFieldName).(string)
+	}
+	if d.HasChange(auditLogDescriptionFieldName) {
+		auditLog.Description = d.Get(auditLogDescriptionFieldName).(string)
+	}
```

Comment:
I think organization_id also needs to be checked and change should be prevented/reported here